applications/FluidDynamicsHydraulicsApplication/custom_utilities/hydraulic_fluid_auxiliary_utilities.cpp

```diff
@@ -302,9 +302,141 @@ void HydraulicFluidAuxiliaryUtilities::SetInletFreeSurface(ModelPart &rModelPart
         if (rNode.Is(rSkinFlag)){
             double inlet_dist = rNode.GetValue(rDistanceVariable);
             rNode.FastGetSolutionStepValue(DISTANCE) = inlet_dist;
-            rNode.Fix(DISTANCE);
+            // The distance is fixed in the water nodes since the inlet has velocity.
+            if (inlet_dist<0.0){
+                rNode.Fix(DISTANCE);
+            }
+        }
+    });
+}
+
+void HydraulicFluidAuxiliaryUtilities::TurnOffGravityOnAirElements(ModelPart &rModelPart){
+    block_for_each(rModelPart.Elements(), [](Element &rElement){
+        auto &r_geom = rElement.GetGeometry();
+        Vector distances = ZeroVector(r_geom.PointsNumber());
+        Vector gravity = ZeroVector(3);
+        for (unsigned int i_nodes = 0; i_nodes < r_geom.PointsNumber(); i_nodes++){
+            distances[i_nodes] = r_geom[i_nodes].FastGetSolutionStepValue(DISTANCE);
+            }
+        if (FluidAuxiliaryUtilities::IsPositive(distances)){
+            for (unsigned int i_nodes = 0; i_nodes < r_geom.PointsNumber(); i_nodes++)
+            {
+                r_geom[i_nodes].FastGetSolutionStepValue(BODY_FORCE) = gravity;
+            }
+        }
+    });
+}
+
+void HydraulicFluidAuxiliaryUtilities::FixCornerNodeVelocity(ModelPart &rModelPart, double MaximumAngle)
+{
+    // Obtain for each condition each neighbor condition
+    block_for_each(rModelPart.Nodes(), [&](NodeType &rNode){ rNode.SetValue(AUX_INDEX, 0); });
+    ModelPart::ConditionsContainerType &r_cond = rModelPart.Conditions();
+    double angle_corner_rad = MaximumAngle / 180.0 * 3.14159; //
+    double acceptable_cos = cos(angle_corner_rad);
+    double max_height = block_for_each<MaxReduction<double>>(rModelPart.Nodes(), [&](NodeType &rNode){return rNode.Z();});
+
+    block_for_each(rModelPart.Nodes(),[&](NodeType &rNode){
+       if(rNode.Z() >=max_height){
+         rNode.Set(MARKER,true);
+       }
+    });
+    for (ModelPart::ConditionsContainerType::iterator cond_it = r_cond.begin(); cond_it != r_cond.end(); cond_it++)
+    {
+        // reference for area normal of the face
+        const array_1d<double, 3> &face_normal = cond_it->GetValue(NORMAL);
+        double An = norm_2(face_normal);
+        const GlobalPointersVector<Condition> &neighb = cond_it->GetValue(NEIGHBOUR_CONDITIONS);
+        const GeometryType &r_geom = cond_it->GetGeometry();
+        auto edgelist = r_geom.GenerateEdges();
+        for (unsigned int c_itr = 0; c_itr < neighb.size(); c_itr++){
+            auto rCondition = neighb[c_itr];
+            const array_1d<double, 3> &face_neig_normal = rCondition.GetValue(NORMAL);
+            double cos_angle = 1 / (An * norm_2(face_neig_normal)) * inner_prod(face_normal, face_neig_normal);
+            auto &r_geom_neig = rCondition.GetGeometry();
+            auto edgelist_neig = r_geom_neig.GenerateEdges();
+            array_1d<double, 2> ids_neig;
+            array_1d<double, 2> ids_cond;
```

Comment:
```suggestion
            array_1d<double, 2> ids_neig;
            array_1d<double, 2> ids_cond;
```
Allocate these two outside the loop.